modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -149,28 +150,75 @@ def __unmasked_losses(
             losses /= mask_mean
 
         return losses
-
-    def __min_snr_weight(
+    
+    def calculate_loss(
             self,
-            timesteps: Tensor,
-            gamma: float,
-            v_prediction: bool,
-            device: torch.device
-    ):
+            model,
+            batch: dict,
+            data: dict,
+            config: TrainConfig,
+    ) -> Tensor:
+        return self._diffusion_losses(
+            batch=batch,
+            data=data,
+            config=config,
+            train_device=self.train_device,
+            betas=model.noise_scheduler.betas.to(device=self.train_device),
+        ).mean()
+    
+    def __snr(self, timesteps: Tensor, device: torch.device):
         if self.__coefficients:
             all_snr = (self.__coefficients.sqrt_alphas_cumprod /
                        self.__coefficients.sqrt_one_minus_alphas_cumprod) ** 2
             all_snr.to(device)
             snr = all_snr[timesteps]
         else:
             alphas_cumprod = self.__alphas_cumprod_fun(timesteps, 1)
-            snr = ((alphas_cumprod ** 0.5) / (1.0 - alphas_cumprod) ** 0.5) ** 2
+            snr = alphas_cumprod / (1.0 - alphas_cumprod)
+        
+        return snr
+        
+
+    def __min_snr_weight(
+            self,
+            timesteps: Tensor,
+            gamma: float,
+            v_prediction: bool,
+            device: torch.device
+    ) -> Tensor:
+        snr = self.__snr(timesteps, device)
         min_snr_gamma = torch.minimum(snr, torch.full_like(snr, gamma))
         # Denominator of the snr_weight increased by 1 if v-prediction is being used.
         if v_prediction:
-            snr += 1
-        snr_weight = torch.div(min_snr_gamma, snr).float().to(device)
+            snr += 1.0
+        snr_weight = (min_snr_gamma / snr).to(device)
         return snr_weight
+    
+    def __debiased_estimation_weight(
+        self,
+        timesteps: Tensor,
+        v_prediction: bool,
+        device: torch.device
+    ) -> Tensor:
+        snr = self.__snr(timesteps, device)
+        weight = snr
+        torch.clip(weight, min=1.0e-3, max=1.0e3, out=weight)
+        if v_prediction:
+            weight += 1.0
+        torch.rsqrt(weight, out=weight)
+        return weight
+    
+    def __p2_loss_weight(
+        self,
+        timesteps: Tensor,
+        gamma: float,
+        v_prediction: bool,
+        device: torch.device,
+    ) -> Tensor:
+        snr = self.__snr(timesteps, device)
+        if v_prediction:
```

Comment:
More a note for myself here -- I need to go to the P2 paper to double check this. It makes sense but I'm not sure if P2 paper took velocity prediction (or even prediction type at all) into account or not. If it's independent of prediction type we shouldn't have this. Let's double check before merging.